File: day005_math_logic/related_natural_numbers_sum/solution.cpp
#include <iostream>

using namespace std;

// O(1) Time, O(1) Space - Optimal
long long sumNaturalNumbers_optimal(long long N) {
    // Use the Arithmetic Series Formula: N * (N + 1) / 2
    return N * (N + 1) / 2;
}

// O(N) Time, O(1) Space - Brute Force
long long sumNaturalNumbers_brute(long long N) {
    long long sum = 0;
    for (long long i = 1; i <= N; ++i) {
        sum += i;
    }
    return sum;
}

void solve_natural_numbers_sum() {
    long long N;
    if (!(cin >> N)) return;

    // Use optimal O(1) approach
    cout << sumNaturalNumbers_optimal(N) << endl;
}

int main() {
    solve_natural_numbers_sum();
    return 0;
}

File: day005_math_logic/related_natural_numbers_sum/solution.java
import java.util.Scanner;

class Solution {

    // O(1) Time, O(1) Space - Optimal
    public static long sumNaturalNumbers_optimal(long N) {
        // Use the Arithmetic Series Formula: N * (N + 1) / 2
        return N * (N + 1) / 2;
    }

    // O(N) Time, O(1) Space - Brute Force
    public static long sumNaturalNumbers_brute(long N) {
        long sum = 0;
        for (long i = 1; i <= N; i++) {
            sum += i;
        }
        return sum;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLong()) return;
        long N = scanner.nextLong();
        
        // Use optimal O(1) approach
        System.out.println(sumNaturalNumbers_optimal(N));
        scanner.close();
    }
}